lib/oxigraph/src/storage/memory.rs

```diff
@@ -5,75 +5,239 @@ use crate::storage::binary_encoder::{
     write_posg_quad, write_spo_quad, write_spog_quad, QuadEncoding, WRITTEN_TERM_MAX_SIZE,
 };
 pub use crate::storage::error::StorageError;
-use crate::storage::numeric_encoder::{insert_term, EncodedQuad, EncodedTerm, StrHash, StrLookup};
+use crate::storage::numeric_encoder::{
+    insert_term, Decoder, EncodedQuad, EncodedTerm, StrHash, StrLookup,
+};
 use crate::storage::CorruptionError;
+use dashmap::DashMap;
 use oxrdf::Quad;
-use std::cell::{BorrowMutError, Ref, RefCell};
-use std::collections::{BTreeSet, HashMap, HashSet};
+use std::collections::{BTreeMap, HashMap};
 use std::error::Error;
-use std::mem::transmute;
-use std::ops::Deref;
-use std::rc::{Rc, Weak};
-use std::sync::{Arc, PoisonError, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::hash::{BuildHasherDefault, Hasher};
+use std::iter::Chain;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex, RwLock};
 
 const BULK_LOAD_BATCH_SIZE: u64 = 100_000;
 
-/// Low level storage primitives
+pub type MemoryChainedQuadIterator =
+    Chain<std::vec::IntoIter<EncodedQuad>, std::vec::IntoIter<EncodedQuad>>;
+pub type MemoryDecodingGraphIterator = std::vec::IntoIter<EncodedTerm>;
+
+/// In-memory storage working with MVCC
+///
+/// Each quad and graph name is annotated by a version range, allowing to read old versions while updates are applied.
+/// To simplify the implementation a single write transaction is currently allowed. This restriction should be lifted in the future.
 #[derive(Clone)]
 pub struct MemoryStorage {
     content: Arc<RwLock<Content>>,
-    id2str: Arc<RwLock<HashMap<StrHash, String>>>,
+    id2str: Arc<DashMap<StrHash, String, BuildHasherDefault<StrHashHasher>>>,
```

Comment:
Not necessarily better or worse, but the `concread` crate might a good fit for the desired MVCC semanitcs here. Not sure if you really can rely on their versioning though as need to keep multiple maps in sync.

Reply:
Thank you! I had a look at it. Sadly, it is under MPLv2 and Oxigraph is Apache 2/MIT. I did not wanted to introduce MPLv2 dependencies.